### ISSUE TITLE
0.55.x (latest: 0.55.4)

### ISSUE THREAD

**Issue by grabbou** [react-native-community/react-native-releases#2]
Latest release: 0.55.4

**Comment by grabbou** [comment]
Status: Ci tests failing 
Date: 06/03/2018
Next update: 06/03/2018
Expected status: CI tests passing
Note: There seems to be an unexpected error due to Metro bundler version mismatch. Teams are aware of the problem. I'll be following up later today.

**Comment by grabbou** [comment]
Status: CI tests failing
Date: 07/03/2018
Next update: 07/03/2018
Expected status: CI tests passing
Note: Apologies for radio silence yesterday, but didn't have enough time to check with Metro team what's going on. I'll attempt fixing the issue today, we hopefully go green really soon.

**Comment by grabbou** [comment]
Status: CI tests failing
Date: 09/03/2018
Next update: 09/03/2018
Expected status: CI tests passing
Note: CI is failing because of a really odd issue (anyone has an idea?) Basically, commit on master passes, whereas same commit breaks on release branch. I have few ideas but waiting for Hector from FB team to check this internally. I can't promise it's going to get fixed today, but I'll definitely update you later in the day.

**Comment by grabbou** [comment]
Yeah, my guess was also that the cache on Circle CI 2.0 has wrong keys (fallback ones) which fails for the first time on any branch where there's no specific cache.

Here's failing test: https://circleci.com/gh/facebook/react-native/36435

**Comment by motiz88** [comment]
Looks like a recurrence of https://github.com/facebook/react-native/issues/17742. (@grabbou do you want me to move this elsewhere?)

## The failure

The `test_ios` Circle CI build is failing with:

```
Error: Unable to resolve module `metro/src/lib/bundle-modules/HMRClient`
  from `/Users/distiller/react-native/Libraries/Utilities/HMRClient.js`:
  Module does not exist in the module map
```

## The state of the filesystem

The actual build filesystem on CircleCI (as cached at the end of https://circleci.com/gh/facebook/react-native/36269) seems to include both `HMRClient.js` and `HmrClient.js` in the requested location:

```
...
./node_modules/metro/src/lib/bundle-modules/MetroClient.js
./node_modules/metro/src/lib/bundle-modules/HMRClient.js.flow
./node_modules/metro/src/lib/bundle-modules/MetroClient.js.flow
./node_modules/metro/src/lib/bundle-modules/asyncRequire.js.flow
./node_modules/metro/src/lib/bundle-modules/asyncRequire.js
./node_modules/metro/src/lib/bundle-modules/HMRClient.js
./node_modules/metro/src/lib/bundle-modules/HmrClient.js
./node_modules/metro/src/lib/bundle-modules/BUCK
./node_modules/metro/src/lib/bundle-modules/HmrClient.js.flow
...
```

## The potential fix

In https://github.com/facebook/react-native/issues/17742 it was pointed out that https://github.com/facebook/react-native/commit/e048bf1a8ae8a0282dca6e5e0b6bf4f010446726 (requiring `HmrClient` instead of `HMRClient`) fixed things. My hypothesis as to why `master` is working is that it has a cache from a time when `HmrClient` was requested, which a clean build cannot reconstruct.

This is likely to be an unreliable fix in the long run; the problem could be down to nondeterministic behaviour in `metro` with filenames that are only distinguishable by letter case, e.g. relying on `readdir` sort order where it is not actually guaranteed to be consistent.

However, based on circumstantial evidence, I believe that changing https://github.com/facebook/react-native/blob/26684cf3adf4094eb6c405d345a75bf8c7c0bf88/Libraries/Utilities/HMRClient.js#L16 to require `HmrClient` will fix the build for the time being.

**Comment by grabbou** [comment]
Thanks for digging into that! I have applied the suggested fix and will try
to see whether it makes the build green. It worries me that the referenced
issue says that it has been fixed in `master` which is clearly not true.

I'll pass this thread to Facebook team to see what's their idea.

Thank you! You saved me few hours of digging :)

**Comment by hramos** [comment]
Amazing. Thanks for looking into this. It makes total sense and explains the flaky test behavior.

**Comment by hramos** [comment]
I'm landing a fix. We were using a stale `node_modules` cache. While the same cache is used by all jobs in the `tests` workflow, the issue would only surface on workflows running on case insensitive OS X machines. 

i.e. when the cache was restored and the workspace attached unto our Ubuntu containers, we'd end up with the stale `HmrClient.js` as well as the correct `HMRClient.js` file, and we'd happily move along.

On OS X, however, upon attaching the workspace we'd end up with a single `HmrClient.js` file which is not what we expect.

Note that this all comes from an issue we had with Metro. The release that preceded 0.25 was published to npm with the wrong case, i.e. it had `HmrClient.js` in the published package, even though the repo itself was tracking `HMRClient.js`. This happened due to a mismatch of case sensitiveness between the machines hosting our repo and the developer's machine that published the package.

As part of my fix, I added a regression test to ensure this doesn't happen again.

**Comment by hramos** [comment]
Note that my fix is landing on master, we might need to cherry-pick it unto 0.55-stable.

**Comment by grabbou** [comment]
Great, I just took the latest `master` to 0.55-branch, attempting release today. 

**Status**: CI tests failing
**Date**: 12/03/2018
**Next update**: 12/03/2018
**Expected status**: RC published
**Note**: Now that cache issues are gone, I expect this to be a smooth process.

**Comment by grabbou** [comment]
Status: RC published
Date: 13/03/2018

**Comment by hramos** [comment]
CI tests are currently failing due to a BUCK issue. Introduced in https://github.com/facebook/react-native/commit/3a2bdf5c50d48305d1538256a816420a6ebcaa48. Fix landing soon.

**Comment by hramos** [comment]
^ The above comment is for master, and I realize this may not be affecting 0.55-RC as the problematic commit is not in the RC.

Do you care about CI test failures on master? I can see this being of interest to you when the RC cutoff date is near, but probably less so mid-cycle (this failure will be long fixed by the time 0.56-RC is cut).

**Comment by grabbou** [comment]
Yeah, would be good to be aware of the status anyway. That way, I can also help fixing the CI when other team members are busy, which should improve the PR review process. I am thinking whether we should have a separate issue for tracking master status.

FYI, I am going to cherry-pick a commit requested recently to add support for context and cut new release candidate today.

https://github.com/facebook/react-native/commit/4f8328bf2fb71b104baa70178a1e1f9636773c6a

Status update to follow in few minutes.

**Comment by grabbou** [comment]
Also, going to cherry-pick this to fix issues with `link` after recent rewrite: https://github.com/facebook/react-native/pull/18349

**Comment by grabbou** [comment]
Status: CI tests pending
Date: 20/03/2018
Next update: 20-21/03/2018
Expected status: RC published

**Comment by joshyhargreaves** [comment]
@grabbou I think this PR would be a good one to cherry-pick. Not landed yet but would generally be a good idea to https://github.com/facebook/react-native/pull/18443

**Comment by grabbou** [comment]
Status: RC published
Date: 21/03/2018

**Comment by grabbou** [comment]
**Status**: CI tests pending
**Date**: 28/03/2018
**Next update:** 28/03/2018
**Expected status**: RC published (0.55.0-rc.2)

**Comment by grabbou** [comment]
Status: CI tests pending
Date: 04/04/2018
Next update: 03/04/2018
Expected status: Stable published (0.55.0)
--
Welcome after Easter break :)

**Comment by grabbou** [comment]
**Status**: Stable released
**Date**: 04/04/2018

**Comment by grabbou** [comment]
Note: Current version (0.55.0) ships with alpha version of React. This is a mistake. Unfortunately, I can't unpublish the release, so we are waiting for a sync to come from Facebook. Expect 0.55.1 to be published later today / tomorrow morning containing stable version of React.

**Comment by grabbou** [comment]
**Status**: Stable released (0.55.1)
**Date**: 06/04/2018
**Note**: This version fixes several issues related to `run-ios` and updates React to stable version

**Comment by grabbou** [comment]
Status: Stable published (0.55.2)
Date: 09/04/2018

**Comment by grabbou** [comment]
**Status**: CI tests pending
**Date**: 17/04/2018
**Expected status**: Stable published (0.55.3)

**Comment by grabbou** [comment]
**Status**: Stable published (0.55.3)
**Date**: 17/04/2018

**Comment by grabbou** [comment]
**Status**: Stable published (0.55.4)
**Date**: 09/05/2018
